### ISSUE TITLE
strip_compile_prefix covers only the final save: HF checkpoints under use_fsdp2_compile still load as a dead adapter

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#351]
Follow-up to #335, measured at 70B on 8x H100. Full record: `benchmarks/gate-h100-validation.md`, STEP 28.

**The #335 repair works, and this issue is not a regression report.** After a 100-step `use_fsdp2_compile: true` run on Qwen2.5-72B, the log prints `Normalised 320 adapter keys saved through torch.compile's wrapper` and the output root is clean. That is the first demonstration of the repair above 0.5B.

What it does not cover is the checkpoint directories HF's own Trainer writes:

| file | tensors | canonical keys | `_orig_mod.` keys |
|---|---|---|---|
| output root | 320 | **320** | 0 |
| `checkpoint-100` | 320 | **0** | **320** |

`sft.py:1498` calls `strip_compile_prefix(self._output_dir)` after `save_model`, i.e. on the final save only. Every `checkpoint-*` directory keeps torch.compile's prefix on every key.

## Why it matters

`PeftModel.from_pretrained` matches by canonical name, so a `checkpoint-*` directory loads with **0 non-zero `lora_B`** — the original #335 failure shape: a file full of real numbers that loads as zeros, with only a `UserWarning`. It bites anyone who

- resumes from a checkpoint,
- points `--adapter` at a `checkpoint-*` directory,
- or keeps checkpoints as the artifact and discards the run directory.

## Reproducer

Any `use_fsdp2_compile: true` run that writes a checkpoint. Then:

```python
from safetensors.torch import load_file
sd = load_file("<out>/checkpoint-N/adapter_model.safetensors")
print(sum(k.startswith("base_model.") for k in sd), sum(k.startswith("_orig_mod.") for k in sd))
# measured: 0 320
```

## Suggested fix

Normalise inside the save path rather than after the final save — e.g. a `TrainerCallback.on_save` that runs `strip_compile_prefix` on the checkpoint directory, so both the periodic and the final artifacts are canonical.

## Acceptance criteria

- [ ] A `checkpoint-*` written under `use_fsdp2_compile` contains 0 `_orig_mod.` keys
- [ ] Test asserts the trained VALUES survive the rename in a checkpoint dir, not just the key names (same standard #335's repair was held to)
- [ ] Resuming from such a checkpoint restores non-zero `lora_B`

**Comment by MakazhanAlpamys** [comment]
Severity correction, from @wilyan09007's analysis in #380 and verified here.

This issue says the checkpoint loads "with only a `UserWarning`". That is true through one door and not the other, and the difference is the whole weight of the third acceptance criterion:

- `--adapter <checkpoint-N>` (`PeftModel.from_pretrained`) **does** warn. `peft/peft_model.py:584` carries the comment *"Let's warn here since (in contrast to load_adapter) we don't return the load result"*, then `warnings.warn` at `:598`.
- `resume_from_checkpoint` warns **nothing**. `transformers/trainer.py:3015` calls `model.load_adapter(resume_from_checkpoint, active_adapter, is_trainable=True)` and does not assign the return value; `load_adapter` deliberately does not warn, it hands the missing keys back in the load result, and nothing reads it. The unexpected `_orig_mod.` keys are then dropped by `load_state_dict(..., strict=False)` without a word.

So resuming a 70B run from `checkpoint-100` discards every step trained before it and continues from a re-zeroed `lora_B` in total silence — #335's failure shape with its one warning removed, not a milder version of it.

Same mechanism affects `load_best_model_at_end`, which goes through `trainer.py:3099` `model.load_adapter(self.state.best_model_checkpoint, ...)`.

**Comment by MakazhanAlpamys** [comment]
Closed by #380 from @wilyan09007 (merged as [`55f583b`](https://github.com/MakazhanAlpamys/Soup/commit/55f583b)).

`CompilePrefixCallback` now normalises each `checkpoint-*` on HF's `on_save`, as the Trainer writes it. Two details decided the shape, and both are worth knowing before anyone "simplifies" it:

- **Neither call site is redundant.** `on_save` is dispatched only from `_save_checkpoint` and never for the final `save_model`, so the callback covers the periodic checkpoints and the existing call at the end of `train()` covers the run directory. The comment at each site now says so.
- **It attaches in `setup()`, not `train()`.** `HFPushCallback.on_save` uploads `checkpoint-{step}` on the same event and `CallbackHandler` dispatches in insertion order, so a normalisation attached any later would have left `--push-as` publishing the prefixed adapter while keeping the repaired one on local disk. That was the same defect surviving on the one path where it cannot be undone afterwards. @wilyan09007 took this rather than deferring it, which was the right call.

Both call sites are guarded on `args.should_save` — the condition `save_model` itself writes under — rather than `state.is_world_process_zero`. The two diverge under `save_on_each_node`, which Soup does not set today, so this is a guard against drift rather than a live fix.

**`load_best_model_at_end` is repaired by the same change**, since it restores `state.best_model_checkpoint` through the same `load_adapter` call at `trainer.py:3099` and was reloading a dead adapter for the same reason.

The severity correction above — that `resume_from_checkpoint` warns *nothing*, where `--adapter` at least warns — is also @wilyan09007's, from the PR analysis, and it is what made the third acceptance criterion the one carrying the weight here.
